[ChangIkJoong] iP

ChangIkJoong opened the pull request:
First Pull Request as IP assignment's requirements. Hopefully this works!!

ChangIkJoong committed:
Level-0

ChangIkJoong committed:
Added base functionality

Added the base functionality per week 1 + tag to the project.

ChangIkJoong committed:
Implement input management and exit prompt

Created a taskManager() to handle the input prompt from the user, and added a Scanner function for input itself from the user. Using input "bye" from the user will stop the program via the while loop utilizing isRunning boolean.

ChangIkJoong committed:
create add, list functionality

added the functionality to add new prompts to a list, as well as the ability to list items via taskManager().

Future functionality would be to not include blank lines in the item list.

ChangIkJoong committed:
implement marking functionality

From previous input, created a method to handle blank input, and implemented markDone and unmarkDone for tasks in the Bruce program.

Future functionality would be to push for an response if task already has been marked done or undone.

ChangIkJoong committed:
remove mark bug

Quick update to recognize the isCompleted in outputInterface to match the assignemts'.

ChangIkJoong committed:
ABbstract and implement views and new Task classes

Updated the abstraction for a MVC (model view controller) to make it easier to understand and navigate.

Implemented new Task classes and methods to handle these accordingly to Deadlines and Events and reflect in the UI.

ChangIkJoong committed:
Further abstraction and improve code readability

adjustments to abstraction lvls in the code, now it's seperated fully by MVC principle so far to make it easier for future development and abstraction according to instructions, now running instances passed to controller fr MV. Modified certain naming of methods and variables to fit conventions.

Future version, maybe add switch cases /enums to switch between taskManagement states?

ChangIkJoong committed:
improve readability and naming conventions

adjustments to variables and naming conventions.

Future version, maybe add switch cases /enums to switch between taskManagement states?

ChangIkJoong committed:
change visibility for View constants

adjustments to visibility, package-private -> public for static variables.

Future version, maybe add switch cases /enums to switch between taskManagement states?

JeanPerrierIII commented on src/main/java/Controller.java:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    private void handleInput(String inputPrompt) {
+        if (inputPrompt.equalsIgnoreCase("Bye.")) {
+            exitProgram();
+        } else if (inputPrompt.isEmpty()) {
+            view.viewError();
+        } else if (inputPrompt.toLowerCase().startsWith("mark ")) {
+            markTaskDone(inputPrompt);
+        } else if (inputPrompt.toLowerCase().startsWith("unmark ")) {
+            unmarkTaskDone(inputPrompt);
+        } else if (inputPrompt.equalsIgnoreCase("list")) {
+            view.viewTaskList(model.getTasks());
+        } else if (inputPrompt.toLowerCase().startsWith("deadline ")) {
+            addDeadlineTask(inputPrompt);
+        } else if (inputPrompt.toLowerCase().startsWith("event ")) {
+            addEventTask(inputPrompt);
+        } else {
+            addTodoTask(inputPrompt);
+        }
```
maybe try splitting the input prompt into words, then just look at the first word and make switch case

JeanPerrierIII commented on src/main/java/Model.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    public boolean markDone(int index) {
+        if (index >= 0 && index < tasks.size()) {
```
Might be able to extract the range check out for even cleaner code

JeanPerrierIII commented on src/main/java/View.java:
```diff
@@ -0,0 +1,78 @@
+    private String viewTask(Task task) {
+        String resultString = "";
+        if (task instanceof Todo) {
+            resultString = "[T]" + isCompleteStatus(task) + task.getTaskDescription();
+        } else if (task instanceof Deadline) {
+            Deadline deadline = (Deadline) task;
+            resultString = "[D]" + isCompleteStatus(task) + deadline.getTaskDescription() + " (by: " + deadline.getEndDate() + ")";
+        } else if (task instanceof Event) {
+            Event event = (Event) task;
+            resultString = "[E]" + isCompleteStatus(task) + event.getTaskDescription() + " (from: " + event.getStartDate() + " to: " + event.getEndDate() + ")";
+        } else {
+            resultString = isCompleteStatus(task) + task.getTaskDescription();
+        }
```
Might be able to extract these into their own functions within the class for cleaner code, maybe convert the else-ifs to a switch case as well

JeanPerrierIII commented on src/main/java/Model.java:
overall really like how clean the class is

JeanPerrierIII reviewed (commented):
Overall great code, nice and easy to read and understand 👍
Some places could possible be neatened up abit but still good.

JianHao24 commented on src/main/java/Controller.java:
```diff
@@ -0,0 +1,97 @@
+        }
+    }
+
+    private void handleInput(String inputPrompt) {
```
Perhaps a switch case would be better than using if and else!

JianHao24 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,32 @@
+    private String taskDescription;
+    private boolean isDone;
+    private int taskId;
+    private static int counter = 1;
```
can have a better name than counter like taskCounter which is more specific and non generic

ChangIkJoong replied:
Nice catch, thank you! :D

Exceptional-Khoi commented on src/main/java/TaskView.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    public void viewExit() {
+        System.out.println("Bye.");
```
You can use System.exit() to completely stop the running program

Exceptional-Khoi commented on src/main/java/Todo.java:
```diff
@@ -0,0 +1,5 @@
+    public Todo(String taskDescription) {
+        super(taskDescription);
+    }
+}
```
The program is well-orgranized. You have many classes, but each class has needed function and does its job great. It creates a good feeling for readers and reviewers, because we don't have to read everything in "Bruce"

Exceptional-Khoi commented on src/main/java/Bruce.java:
```diff
@@ -4,83 +4,102 @@
 
-    public static void markDone(String prompt) {
+    private static void markDone(String prompt) {
         String[] parts = prompt.split(" ");
```
split(" ") doesn't work for the case like: "read *two spaces* book". you can use trim()

Exceptional-Khoi commented on src/main/java/Bruce.java:
```diff
@@ -4,83 +4,102 @@
+            newTodoTask(inputPrompt);
         }
     }
 
```
Bruce class is a little too long. If you add more functions in the future, it'll be even longer. You can move some functions to TaskView class

JianHao24 commented on src/main/java/View.java:
```diff
@@ -0,0 +1,78 @@
+public class View {
+    public static final String INLINE_TEXT_LINES = "____________________________________________________________";
```
Constants are set in SCREAMING_SNAKE_CASE which follows convention. Good Job!

JianHao24 commented on src/main/java/View.java:
```diff
@@ -0,0 +1,78 @@
+        System.out.println(" Now you have " + total + " tasks in the list.");
+    }
+
+    public void viewExit() {
```
Instead of hardcoding "Bye." or "Error. Please try again.", instead define them as constants at the top like (private static final String DEFAULT_ERROR_MESSAGE = "Error. Please try again."; ). This would increase the readability of your code and reduce redudancy when printing the same error message.

ChangIkJoong replied:
As of week 4, we had not yet discussed exceptions and error throwings, this is why I hard coded it in order to not jump ahead, but thank you for great feedback and will implement this in further iteration! :D

JianHao24 commented on src/main/java/Model.java:
```diff
@@ -0,0 +1,39 @@
+        return tasks;
+    }
+
+    public boolean markDone(int index) {
```
boolean can be isDone to better follow Java conventionand named to sound like a boolean instead of markDone

ChangIkJoong replied:
The line you are reffering to is a method not a boolean variable. According to conventions this should not be a problem. Please correct me if I'm wrong, thank you :) !

ChangIkJoong committed:
Implement PR log comments from peers

Implemented bug fix for program still printing "Event" or "deadline" in the task descriptions, changed Controller to switch cases (lamda) for easier readability, changed variable names (taskcounter). added range check as seperate method for easier readability in model.

ChangIkJoong committed:
add A-packaging, implement exception handling and code-A-standard

Implemented bug fixes, error handling for all task versions so far, also logic for this. Future work needs to be done in the error handling, also abstraction and interfacing of the task model. Right now 1 parent, 3 children.

ChangIkJoong committed:
Merge Level-5 increment

ChangIkJoong committed:
add A-packaging and refactoring of Task Variants

Changed the packaging conventions for the project to fit better A-Code quality.

ChangIkJoong committed:
Merge branch-A-Packages increment

JoanneHing commented on src/main/java/UI/View.java:
```diff
@@ -0,0 +1,99 @@
+        System.out.println("  bye.                                  : Exits the chatbot.");
+        System.out.println("  help                                  : Shows this instruction list.");
+    }
+}
```
Great job setting up the view architecture for iP. It's a good application of what covered in the lecture.

JoanneHing replied:
Missed out on this, instead using lots of `get` method to get the detail of the tasks, maybe you can consider overwritting the `toString()` method of the task classes better?

JoanneHing commented on src/main/java/Bruce/Controller.java:
```diff
@@ -0,0 +1,140 @@
+public class Controller {
+    private final Model model;
+    private final View view;
+    private boolean isRunning = true;
```
Good job in naming the boolean following the coding standards.

JoanneHing commented on src/main/java/Bruce/Controller.java:
```diff
@@ -0,0 +1,140 @@
+    }
+
+    public void runProgram() {
+        View.greetUser();
+        while (isRunning) {
+            String inputPrompt = scanner.nextLine();
+            try {
+                handleInput(inputPrompt);
+            } catch (NumberFormatException e) {
+                view.viewError("Task number must be a whole number. Example: \"mark 3\"");
+            } catch (IndexOutOfBoundsException e) {
+                view.viewError("Missing task number. Example: \"mark 3\"");
+            } catch (DukeException e) {
+                view.viewError(e.getMessage());
+            } catch (Exception e) {
+                view.viewError("Unexpected error: " + e.getMessage());
+            } finally {
+                View.printLine();
+            }
+        }
+    }
```
Good to see that you include the finally block in a try-catch block, its a good practice!

JoanneHing commented on src/main/java/Bruce/Controller.java:
```diff
@@ -0,0 +1,140 @@
+    }
+
+    private void handleInput(String inputPrompt) throws DukeException {
+        String command = inputPrompt.trim().split(" ")[0].toLowerCase();
+
+        switch (command) {
+        case "bye." -> exitProgram();
+        case "mark" -> markTaskDone(inputPrompt);
+        case "unmark" -> unmarkTaskDone(inputPrompt);
+        case "list" -> view.viewTaskList(model.getTasks());
+        case "deadline" -> addDeadlineTask(inputPrompt);
+        case "event" -> addEventTask(inputPrompt);
+        case "todo" -> addTodoTask(inputPrompt);
+        case "help" -> View.showInstructions();
+        default -> throw new UnknownInputException(command);
```
It is allowed to use lambda blocks, but remember to indent the `case` blocks, as the coding standard would suggest lambda `switch-case` to have indented `case`. May review the coding standard page for related reference for "Lambda-style switch statements/expressions can have indented case blocks"

JoanneHing commented on src/main/java/Model/Model.java:
```diff
@@ -0,0 +1,45 @@
+    private boolean rangeCheck(int index) {
+        return (index >= 0 && index < tasks.size());
+    }
+}
```
Good job on implementing the Model architecture too.

JoanneHing commented on src/main/java/Model/TaskVariants/Deadline.java:
```diff
@@ -0,0 +1,20 @@
+package Model.TaskVariants;
+
+public class Deadline extends Task {
+    protected String endDate;
```
May I clarify why you decided to use `protected` instead of `private` variables? Would you have subclass that would need to use this variable?

JoanneHing commented on src/main/java/Model/TaskVariants/Event.java:
```diff
@@ -0,0 +1,30 @@
+package Model.TaskVariants;
+
+public class Event extends Task {
+    protected String startDate;
+    protected String endDate;
```
Same as the comment in `Deadline.java`

JoanneHing reviewed (commented):
Overall well done! Good job to implement MVC architecture that is taught in the lecture too.

ChangIkJoong committed:
add input and output handler class

Added a CodecConverter for both encoding and decoding the txt file, as well as created a dataHandler to communicate between controller and the codecConverter. Encapsulated both the model and datahandler with a singleton pattern to just make sure there's only a single instance of them, might not be needed fully at the moment however still a good precaution for future updates.

ChangIkJoong committed:
Merge Level-7 increment

ChangIkJoong committed:
make .txt update every iteration

It now adds and supports continous updating every time a new item is added.

ChangIkJoong committed:
update to code standard

changed som imports order such as alphabetical and sorted.

ChangIkJoong committed:
add Jar Manifest

according to lecture, adding the .jar manifest to push the correct way for jar file.

ChangIkJoong committed:
Update project name from Duke to Bruce in README

Updated the readme to better fit with current project.

ChangIkJoong committed:
add Java Documentation

Added java documentation inline code for both classes and methods, all public covered.

ChangIkJoong committed:
add find functionality

Added the method findTasks to be able to provide a keyword and find all corresponding tasks including the keyword in their description.

ChangIkJoong committed:
Merge branch-A-JavaDoc

ChangIkJoong committed:
Merge branch-Level-9

ChangIkJoong committed:
debug find functionality

Re-Added the method findTasks to be able to provide a keyword and find all corresponding tasks including the keyword in their description.

ChangIkJoong committed:
clean up code

cleaned up some code and made it easier to understand.

ChangIkJoong committed:
Merge pull request #1 from ChangIkJoong/branch-Code-Improvement

PR clean up code

ChangIkJoong committed:
add UserGuide and image directory

Added the README.md userguide in the docs folder, as well as a sub-folder with image for the .md file.

ChangIkJoong committed:
bug-fix read me

the image was not showing properly, hopefully fixed now

ChangIkJoong committed:
bug-fix screenshot in README

wrong name in previous commit, now it should work.
